website/blog/2025-08-30-mlflow-prompt-evaluate/index.md

```diff
@@ -0,0 +1,503 @@
+
+    if files := os.listdir(directory):
+        selected_files = random.sample(files, k=n)
+        return [file.rsplit(".", 1)[0] for file in selected_files]
+
+    return []
+
+def get_image(
+    file_name: str, directory: str = IMAGES_DIRECTORY) -> bytes:
+    file_name += ".png" if not file_name.endswith(".png") else file_name
+    path = os.path.join(directory, file_name)
+
+    return _compress_image(path)
+
+def encode_image_as_base64(image_bytes: bytes) -> str:
+    """Convert image bytes to base64 string."""
+    return base64.b64encode(image_bytes).decode("utf-8")
+
+def _compress_image(file_path: str, quality: int = 40, max_size: tuple[int, int] = (1000, 1000)) -> bytes:
+    """Compresses an image by resizing and converting to JPEG with given quality."""
+    with im.open(file_path) as img:
+        img = img.convert("RGB")
+        img.thumbnail(max_size)
+        buf = BytesIO()
+        img.save(buf, format="JPEG", quality=quality)
+
+        return buf.getvalue()
+
+def clean_keys(key: str) -> str:
+    """Remove unwanted characters and convert to upper case."""
+    key = re.sub(r'[^\w\s\-_]', '', key).upper()
+    return key
+
+def normalize_json_keys(json_obj: Any) -> Any:
+
+    if isinstance(json_obj, dict):
+        return {clean_keys(k): normalize_json_keys(v) for k, v in json_obj.items()}
+    elif isinstance(json_obj, list) or isinstance(json_obj, pd.Series):
+        return [normalize_json_keys(item) for item in json_obj]
+    else:
+        return json_obj
+
+```
+
+Let’s take a moment to break down the `_extract_qa_pairs` function:
+
+1. We create a look-up list of tuples for preserving duplicates - qa_pairs
+2. We identify "question" items that have linked answer pair(s) in the form of (q_id, a_id)
+3. We construct question-answer pairs by joining the individual words
+
+We can then call `get_json` to fetch the OCR-identified question/answer structure based on the form image data.
+
+Similarly, `get_image` can serve as a utils function that reads an image file from a specified path and resizes or converts it into a JPEG with a defined quality setting. To normalise and clean up any JSON key strings, we can use `normalise_json_keys`.
+
+```python
+from IPython.display import Image, display
+import random
+
+random_file = get_random_files()
+image_bytes = get_image(random_file[0])
+get_json(random_file)
+
+```
+
+### 3. MLflow Tracking, Autolog and Tracing
+
+Before we make a call to OpenAI, we need to set up [MLflow Tracking](https://mlflow.org/docs/latest/ml/tracking) to ensure every experiment, prompt, and result is recorded and traceable.
+
+We will also enable [MLflow Autolog](https://mlflow.org/docs/latest/ml/tracking/autolog) to simplify the tracking process by reducing the need for manual log statements. Metrics, params, artifacts, and other useful information are logged automatically.
+
+```python
+import mlflow
+
+mlflow.set_tracking_uri("./mlruns")
```

Comment:
Could we use sqllite instead of filestore for the tracking backend? We're working on deprecating filestore due to its inability to support many new features we're building.